lms/static/scripts/frontend_apps/utils/api.ts

```diff
@@ -209,3 +212,75 @@ export function useAPIFetch<T = unknown>(
   const paramStr = recordToQueryString(params ?? {});
   return useFetch(path ? `${path}${paramStr}` : null, fetcher);
 }
+
+export type PaginatedFetchResult<T> = Omit<FetchResult<T>, 'mutate'> & {
+  /** Determines if currently loading the first page */
+  isInitiallyLoading: boolean;
+  loadNextPage: () => void;
+};
+
+/**
+ * Hook that fetches paginated data using authenticated API requests.
+ * It expects the result to include pagination info in a `pagination` prop.
+ *
+ * @param prop - Prop containing the list of results to collect.
+ * @param path - Path for API call, or null if there is nothing to fetch
+ * @param [params] - Query params for API call
+ */
+export function usePaginatedAPIFetch<
+  Prop extends string,
+  ListType extends any[],
+  FetchType extends { [P in Prop]: ListType } & { pagination: Pagination },
+>(
+  prop: Prop,
+  path: string | null,
+  params?: Record<string, string | string[] | undefined>,
+): PaginatedFetchResult<ListType> {
+  const [nextPageURL, setNextPageURL] = useState<string>();
+  const [currentList, setCurrentList] = useState<ListType>();
+
+  const fetchResult = useAPIFetch<FetchType>(
+    nextPageURL ?? path,
+    !nextPageURL ? params : undefined,
+  );
+
+  const loadNextPage = useCallback(() => {
+    if (!fetchResult.isLoading && fetchResult.data?.pagination.next) {
+      setNextPageURL(fetchResult.data.pagination.next);
+    }
+  }, [fetchResult.data?.pagination.next, fetchResult.isLoading]);
+
+  // Every time data is loaded, append the result to the list
+  useEffect(() => {
+    setCurrentList(prev => {
+      if (!fetchResult.data) {
+        return prev;
+      }
+
+      return [...(prev ?? []), ...fetchResult.data[prop]] as ListType;
+    });
+  }, [prop, fetchResult.data]);
+
+  // Every time params change, discard previous pages and start from scratch
+  const prevParamsRef = useRef<typeof params>(params);
+  useEffect(() => {
+    const prevParams = JSON.stringify(prevParamsRef.current);
+    const newParams = JSON.stringify(params);
+
+    if (prevParams !== newParams) {
+      setCurrentList(undefined);
+      setNextPageURL(undefined);
+    }
+
+    prevParamsRef.current = params;
+  }, [params]);
+
+  return {
+    isLoading: fetchResult.isLoading,
+    isInitiallyLoading: fetchResult.isLoading && currentList === undefined,
+    data: currentList ?? null,
+    error: fetchResult.error,
+    retry: fetchResult.retry,
```

Comment:
`error` will represent the error loading the last page that was attempted, and `retry` will only retry that page.

I think that's reasonable, and probably what one would expect.